File: chapter10/primes.exs
defmodule Primes do
  def prime_numbers(upper_limit) do
    prime_numbers(Enum.to_list(2..upper_limit), [])
  end

  defp prime_numbers([], acc) do
    Enum.reverse(acc)
  end

  defp prime_numbers([p|rest], acc) do
    prime_numbers((for r <- rest, rem(r, p) != 0, do: r), [p|acc])
  end
end

defmodule Primes2 do
  def prime_numbers([]), do: []
  def prime_numbers([p|rest]), do: [p|prime_numbers(for r <- rest, rem(r, p) != 0, do: r)]
end

File: chapter10/primes.erl
-module(primes).
-export([prime_numbers/1]).

prime_numbers(UpperLimit) ->
  prime_numbers(lists:seq(2, UpperLimit), []).

prime_numbers([], Acc) ->
  lists:reverse(Acc);
prime_numbers([P|Rest], Acc) ->
  prime_numbers([R || R <- Rest, R rem P /= 0], [P|Acc]).
